### PR TITLE
SDK-554: Add support of S3 Object Lock for ECS 3.6.2.

### PR DESCRIPTION
**Pull request by xiaoxin-ren**

### CODE REVIEW

File: src/test/java/com/emc/object/s3/S3JerseyClientTest.java [hunk 1]
```diff
@@ -199,6 +210,228 @@ public void testCreateEncryptedBucket() {
...
+    @Test
+    public void testSetObjectLockConfiguration() {
+        Assume.assumeTrue("Skip Object Lock related tests for non IAM user.", testIAM);
+
+        String bucketName = getTestBucket();
+        ObjectLockConfiguration objectLockConfig = new ObjectLockConfiguration().withObjectLockEnabled(ObjectLockConfiguration.ObjectLockEnabled.Enabled);
+        DefaultRetention defaultRetention = new DefaultRetention().withMode(ObjectLockRetentionMode.GOVERNANCE).withDays(2);
+        objectLockConfig.setRule(new ObjectLockRule().withDefaultRetention(defaultRetention));
+        try {
+            client.setObjectLockConfiguration(bucketName, objectLockConfig);
```
Comment (twincitiesguy) [inline src/test/java/com/emc/object/s3/S3JerseyClientTest.java:245]: add `Assert.fail()` here - if no exception is thrown, that should fail the test, correct?

File: src/test/java/com/emc/object/s3/S3JerseyClientTest.java [hunk 2]
```diff
@@ -199,6 +210,228 @@ public void testCreateEncryptedBucket() {
...
+        client.enableObjectLock(bucketName);
+        ObjectLockLegalHold objectLockLegalHold = new ObjectLockLegalHold().withStatus(ObjectLockLegalHold.Status.ON);
+        PutObjectRequest putObjectRequest = new PutObjectRequest(bucketName, key,"test Delete With LegalHold Not Allowed")
+                .withObjectLockLegalHold(objectLockLegalHold);
+        client.putObject(putObjectRequest);
+        String versionId = client.listVersions(bucketName, key).getVersions().get(0).getVersionId();
+
+        Assert.assertEquals("ON", client.getObjectLegalHold(new GetObjectLegalHoldRequest(bucketName,key).withVersionId(versionId)).getStatus());
+        try {
+            client.deleteVersion(bucketName, key, versionId);
```
Comment (twincitiesguy) [inline src/test/java/com/emc/object/s3/S3JerseyClientTest.java:277]: add `Assert.fail` here - if no exception is thrown, that should fail the test

File: src/test/java/com/emc/object/s3/S3JerseyClientTest.java [hunk 3]
```diff
@@ -199,6 +210,228 @@ public void testCreateEncryptedBucket() {
...
+                .withRetainUntilDate(retentionDate);
+        //Put Retention on Create
+        PutObjectRequest putObjectRequest = new PutObjectRequest(bucketName, key,"test Put Object Retention")
+                .withObjectLockRetention(objectLockRetention);
+        client.putObject(putObjectRequest);
+        String versionId = client.listVersions(bucketName, key).getVersions().get(0).getVersionId();
+        GetObjectRetentionRequest request = new GetObjectRetentionRequest(bucketName,key).withVersionId(versionId);
+        ObjectLockRetention objectLockRetention2 = client.getObjectRetention(request);
+        Assert.assertEquals(objectLockRetention.getMode(), objectLockRetention2.getMode());
+        Assert.assertEquals(0, retentionDate.compareTo(objectLockRetention2.getRetainUntilDate()));
```
Comment (twincitiesguy) [inline src/test/java/com/emc/object/s3/S3JerseyClientTest.java:330]: you can `assertEquals()` the `Date`s directly

**Review by twincitiesguy — CHANGES_REQUESTED** [review]

File: src/main/java/com/emc/object/s3/bean/DefaultRetention.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlType;
+import java.io.Serializable;
+
+@XmlType(propOrder = {"mode", "days", "years"})
+/**
+ * DefaultRetention requires both a mode and a period. The period can be either Days or Years, not both.
+ */
+public class DefaultRetention implements Serializable {
+    private String mode;
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/bean/DefaultRetention.java:12]: This refers to the definition by com.amazonaws:aws-java-sdk-s3 which uses String.
Also see in https://docs.aws.amazon.com/AmazonS3/latest/API/API_DefaultRetention.html

File: src/main/java/com/emc/object/s3/bean/ObjectLockConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
+package com.emc.object.s3.bean;
+
+import javax.xml.bind.annotation.*;
+
+@XmlRootElement(name = "ObjectLockConfiguration")
+@XmlType(propOrder = {"objectLockEnabled", "rule"})
+public class ObjectLockConfiguration {
+    private String objectLockEnabled;
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/bean/ObjectLockConfiguration.java:8]: Just to keep consistent to the definition of implementation by AWS SDK. 
It seems AWS Java SDK defines data type only by following https://docs.aws.amazon.com/AmazonS3/latest/API/API_ObjectLockConfiguration.html, I traced into their SDK and confirmed this.
Technically no problem to define this private member as enum as long as interface is kept compatible to what AWS JavaSDK doc writes: https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/ObjectLockConfiguration.html

File: src/main/java/com/emc/object/s3/request/CreateBucketRequest.java [hunk 6]
```diff
@@ -197,4 +199,13 @@ public CreateBucketRequest withMetadataSearchKeys(List<MetadataSearchKey> metada
         setMetadataSearchKeys(metadataSearchKeys);
         return this;
     }
+
+    public CreateBucketRequest withObjectLockEnabledForBucket(boolean objectLockEnabled) {
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/CreateBucketRequest.java:203]: This confirms definition by AWSJavaSDK, please see:
https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/CreateBucketRequest.html

<html><body>


CreateBucketRequest | withObjectLockEnabledForBucket(boolean objectLockEnabled) Specifies whether you want S3 Object Lock to be enabled for the new bucket.
-- | --



</body>
</html>

File: src/main/java/com/emc/object/s3/request/CreateBucketRequest.java [hunk 7]
```diff
@@ -197,4 +199,13 @@ public CreateBucketRequest withMetadataSearchKeys(List<MetadataSearchKey> metada
...
+        this.objectLockEnabled = objectLockEnabled;
+        return this;
+    }
+
+    public void setObjectLockEnabledForBucket(boolean objectLockEnabled) {
+        withObjectLockEnabledForBucket(objectLockEnabled);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/CreateBucketRequest.java:209]: Good point! Fixed.

File: src/main/java/com/emc/object/s3/request/SetObjectRetentionRequest.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        return versionId;
+    }
+
+    public SetObjectRetentionRequest withVersionId(String versionId) {
+        this.versionId = versionId;
+        return this;
+    }
+
+    public void setVersionId(String versionId) {
+        withVersionId(versionId);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/SetObjectRetentionRequest.java:78]: Fixed, other set/withs are modified too.

File: src/main/java/com/emc/object/s3/request/SetObjectRetentionRequest.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        return bypassGovernanceRetention;
+    }
+
+    public SetObjectRetentionRequest withBypassGovernanceRetention(Boolean bypassGovernanceRetention) {
+        this.bypassGovernanceRetention = bypassGovernanceRetention;
+        return this;
+    }
+
+    public void setBypassGovernanceRetention(Boolean bypassGovernanceRetention) {
+        withBypassGovernanceRetention(bypassGovernanceRetention);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/SetObjectRetentionRequest.java:91]: Fixed

File: src/test/java/com/emc/object/s3/S3JerseyClientTest.java (hunk 1 above)
Comment (xiaoxin-ren) [reply src/test/java/com/emc/object/s3/S3JerseyClientTest.java:245]: Great suggestion! fixed.

File: src/test/java/com/emc/object/s3/S3JerseyClientTest.java (hunk 2 above)
Comment (xiaoxin-ren) [reply src/test/java/com/emc/object/s3/S3JerseyClientTest.java:277]: Fixed.

File: src/test/java/com/emc/object/s3/S3JerseyClientTest.java (hunk 3 above)
Comment (xiaoxin-ren) [reply src/test/java/com/emc/object/s3/S3JerseyClientTest.java:330]: Fixed.

File: src/main/java/com/emc/object/s3/request/CopyObjectRequest.java [hunk 10]
```diff
@@ -83,6 +89,14 @@ public CopyObjectRequest(String sourceBucketName, String sourceKey, String bucke
         }
         if (acl != null) headers.putAll(acl.toHeaders());
         if (cannedAcl != null) RestUtil.putSingle(headers, S3Constants.AMZ_ACL, cannedAcl.getHeaderValue());
+        if (objectLockLegalHold != null) RestUtil.putSingle(headers, S3Constants.AMZ_OBJECT_LOCK_LEGAL_HOLD, objectLockLegalHold.getStatus());
+        if (objectLockRetention != null) {
+            RestUtil.putSingle(headers, S3Constants.AMZ_OBJECT_LOCK_MODE, objectLockRetention.getMode());
+            if (objectLockRetention.getRetainUntilDate() != null) {
+                RestUtil.putSingle(headers, S3Constants.AMZ_OBJECT_LOCK_RETAIN_UNTIL_DATE,
+                        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").format(objectLockRetention.getRetainUntilDate().toInstant().atOffset(ZoneOffset.UTC)));
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/CopyObjectRequest.java]: Extracted these into RestUtil. Please let me know if any better alternative.

[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/CopyObjectRequest.java]: AWS JAVA SDK converts Retention until date in ISO8601 milliseconds format internally and AWS also responses corresponding queries in same format. Although AWS server accepts ISO8601 format without milliseconds, we don't specially care about rest of the supported formats which are not exposed to user in Java SDK.

File: src/main/java/com/emc/object/s3/bean/DefaultRetention.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (twincitiesguy) [reply src/main/java/com/emc/object/s3/bean/DefaultRetention.java:12]: > Valid Values: GOVERNANCE | COMPLIANCE

Any time there is a confined list of values, we should use an enum if we can.  It doesn't really matter what the AWS SDK does - we can use that SDK to test compatibility with ECS's S3 implementation, but this is a client implementation detail we are free to choose.

[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/bean/DefaultRetention.java:12]: OK, will replace those Enum Strings with Enum.

File: src/main/java/com/emc/object/s3/bean/ObjectLockConfiguration.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (twincitiesguy) [reply src/main/java/com/emc/object/s3/bean/ObjectLockConfiguration.java:8]: The SDK interfaces do not need to be compatible - only the API.  If we use an enum here, it does not affect the wire protocol, but it is a best practice in Java to use enums for a confined set of values.

In this case in particular, there is only one value (`Enabled`), so perhaps there is a better implementation that does not expose the `objectLockEnabled` property to the user.

File: src/main/java/com/emc/object/s3/request/CreateBucketRequest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (twincitiesguy) [reply src/main/java/com/emc/object/s3/request/CreateBucketRequest.java:207]: Again, this is a client implementation detail.  We only need to be API-compatible.  There is no requirement to follow the AWS SDK's method style here.  Our SDK already has an established naming convention and we need to be consistent with it.

File: src/main/java/com/emc/object/s3/bean/ObjectLockConfiguration.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/bean/ObjectLockConfiguration.java]: Addressed.

File: src/main/java/com/emc/object/s3/request/CreateBucketRequest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/CreateBucketRequest.java]: addressed.

File: src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Override
+    public Map<String, List<Object>> getHeaders() {
+        Map<String, List<Object>> headers = super.getHeaders();
+        return headers;
+    }
+
+    @Override
+    public Long getContentLength() {
+        return null; // assume buffering
```
[the remark this answers is not in the document]
Comment (twincitiesguy) [reply src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java:33]: please explain why we need to buffer this...

File: src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    @Override
+    public boolean isChunkable() {
+        return false;
```
[the remark this answers is not in the document]
Comment (twincitiesguy) [reply src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java:38]: please explain why this cannot be chunked...

File: src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java:33]: This request is pretty similar to SetObjectAclRequest, so I choose to follow the overrides as what's done inside SetObjectAclRequest. I don't really know which scenario requires this change for SetObjectAclRequest, but feel safer to keep the same here.

File: src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        return this.legalHold;
+    }
+
+    public SetObjectLegalHoldRequest withLegalHold(ObjectLockLegalHold legalHold) {
+        this.legalHold = legalHold;
+        return this;
+    }
+
+    public void setLegalHold(ObjectLockLegalHold legalHold) {
+        this.withLegalHold(legalHold);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java]: fixed.

File: src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        return this.versionId;
+    }
+
+    public SetObjectLegalHoldRequest withVersionId(String versionId) {
+        this.versionId = versionId;
+        return this;
+    }
+
+    public void setVersionId(String versionId) {
+        this.withVersionId(versionId);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java]: fixed

File: src/main/java/com/emc/object/s3/request/SetObjectRetentionRequest.java [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+        return retention;
+    }
+
+    public SetObjectRetentionRequest withRetention(ObjectLockRetention retention) {
+        this.retention = retention;
+        return this;
+    }
+
+    public void setRetention(ObjectLockRetention retention) {
+        withRetention(retention);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/SetObjectRetentionRequest.java]: fixed.

File: src/main/java/com/emc/object/s3/request/SetObjectRetentionRequest.java [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+    public Map<String, String> getQueryParams() {
+        Map<String, String> queryParams = super.getQueryParams();
+        if (versionId != null) queryParams.put("versionId", versionId);
+        return queryParams;
+    }
+
+    @Override
+    public Map<String, List<Object>> getHeaders() {
+        Map<String, List<Object>> headers = super.getHeaders();
+        if (bypassGovernanceRetention != null) RestUtil.putSingle(headers, S3Constants.AMZ_OBJECT_LOCK_BYPASS_GOVERNANCE_RETENTION, bypassGovernanceRetention.toString());
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/SetObjectRetentionRequest.java:31]: will revise the code where uses Boolean.toString() for consistency.

File: src/main/java/com/emc/object/s3/request/SetObjectRetentionRequest.java [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+        return headers;
+    }
+
+    @Override
+    public Long getContentLength() {
+        return null; // assume buffering
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/SetObjectRetentionRequest.java:37]: this override follows SetObjectAclRequest as SetObjectRetentionRequest looks pretty similar. Honestly, I don't really know why it's added inside SetObjectAclRequest. I guess it's needed here.

File: src/main/java/com/emc/object/s3/request/SetObjectRetentionRequest.java [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    @Override
+    public boolean isChunkable() {
+        return false;
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/SetObjectRetentionRequest.java:42]: The same reason by referring to the implementation of SetObjectAclRequest .

File: src/main/java/com/emc/object/s3/bean/ObjectLockRule.java [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+        return defaultRetention;
+    }
+
+    public ObjectLockRule withDefaultRetention(DefaultRetention defaultRetention) {
+        this.defaultRetention = defaultRetention;
+        return this;
+    }
+
+    public void setDefaultRetention(DefaultRetention defaultRetention) {
+        withDefaultRetention(defaultRetention);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/bean/ObjectLockRule.java]: fixed

File: src/main/java/com/emc/object/s3/request/GetObjectLegalHoldRequest.java [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+        return this.versionId;
+    }
+
+    public GetObjectLegalHoldRequest withVersionId(String versionId) {
+        this.versionId = versionId;
+        return this;
+    }
+
+    public void setVersionId(String versionId) {
+        this.withVersionId(versionId);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/GetObjectLegalHoldRequest.java]: fixed

File: src/main/java/com/emc/object/s3/request/GetObjectRetentionRequest.java [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+        return this.versionId;
+    }
+
+    public GetObjectRetentionRequest withVersionId(String versionId) {
+        this.versionId = versionId;
+        return this;
+    }
+
+    public void setVersionId(String versionId) {
+        this.withVersionId(versionId);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/GetObjectRetentionRequest.java:31]: fixed.

File: src/main/java/com/emc/object/s3/bean/ObjectLockRetention.java [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+        this.mode = mode;
+        return this;
+    }
+
+    public ObjectLockRetention withMode(ObjectLockRetentionMode mode) {
+        return withMode(mode.toString());
+    }
+
+    public void setMode(String mode) {
+        withMode(mode);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/bean/ObjectLockRetention.java]: fixed

File: src/main/java/com/emc/object/s3/bean/ObjectLockRetention.java [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+        return retainUntilDate;
+    }
+
+    public ObjectLockRetention withRetainUntilDate(Date retainUntilDate) {
+        this.retainUntilDate = retainUntilDate;
+        return this;
+    }
+
+    public void setRetainUntilDate(Date retainUntilDate) {
+        withRetainUntilDate(retainUntilDate);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/bean/ObjectLockRetention.java]: fixed

File: src/main/java/com/emc/object/s3/bean/ObjectLockRetention.java [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+import javax.xml.bind.annotation.XmlType;
+import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;
+import java.util.Date;
+
+@XmlRootElement(name = "Retention")
+@XmlType(propOrder = {"mode", "retainUntilDate"})
+public class ObjectLockRetention {
+    private String mode;
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/bean/ObjectLockRetention.java]: fixed.

File: src/main/java/com/emc/object/s3/bean/ObjectLockLegalHold.java [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+        return this;
+    }
+
+    public ObjectLockLegalHold withStatus(Status status) {
+        this.status = status.toString();
+        return this;
+    }
+
+    public void setStatus(String status) {
+        withStatus(status);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/bean/ObjectLockLegalHold.java]: fixed.

File: src/main/java/com/emc/object/s3/bean/ObjectLockLegalHold.java [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlEnum;
+import javax.xml.bind.annotation.XmlRootElement;
+import javax.xml.bind.annotation.XmlType;
+
+@XmlRootElement(name = "LegalHold")
+@XmlType(propOrder = {"status"})
+public class ObjectLockLegalHold {
+    private String status;
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/bean/ObjectLockLegalHold.java]: fixed.

File: src/main/java/com/emc/object/s3/bean/ObjectLockConfiguration.java [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+    private ObjectLockRule rule;
+
+    public ObjectLockConfiguration() {
+    }
+
+    @XmlElement(name = "ObjectLockEnabled")
+    public String getObjectLockEnabled() { return this.objectLockEnabled; }
+
+    public void setObjectLockEnabled(String objectLockEnabled) {
+        this.withObjectLockEnabled(objectLockEnabled);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/bean/ObjectLockConfiguration.java]: fixed.

File: src/main/java/com/emc/object/s3/bean/ObjectLockConfiguration.java [hunk 28]
```diff
@@ -0,0 +1,50 @@
...
+
+    public void setObjectLockEnabled(ObjectLockEnabled objectLockEnabled) {
+        this.setObjectLockEnabled(objectLockEnabled.toString());
+    }
+
+    @XmlElement(name = "Rule")
+    public ObjectLockRule getRule() { return this.rule; }
+
+    public void setRule(ObjectLockRule rule) {
+        this.withRule(rule);
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/bean/ObjectLockConfiguration.java]: fixed.

File: src/main/java/com/emc/object/s3/request/InitiateMultipartUploadRequest.java [hunk 29]
```diff
@@ -51,6 +57,14 @@ public InitiateMultipartUploadRequest(String bucketName, String key) {
         if (objectMetadata != null) headers.putAll(objectMetadata.toHeaders());
         if (acl != null) headers.putAll(acl.toHeaders());
         if (cannedAcl != null) RestUtil.putSingle(headers, S3Constants.AMZ_ACL, cannedAcl.getHeaderValue());
+        if (objectLockLegalHold != null) RestUtil.putSingle(headers, S3Constants.AMZ_OBJECT_LOCK_LEGAL_HOLD, objectLockLegalHold.getStatus());
+        if (objectLockRetention != null) {
+            RestUtil.putSingle(headers, S3Constants.AMZ_OBJECT_LOCK_MODE, objectLockRetention.getMode());
+            if (objectLockRetention.getRetainUntilDate() != null) {
+                RestUtil.putSingle(headers, S3Constants.AMZ_OBJECT_LOCK_RETAIN_UNTIL_DATE,
+                        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").format(objectLockRetention.getRetainUntilDate().toInstant().atOffset(ZoneOffset.UTC)));
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/InitiateMultipartUploadRequest.java]: fixed.

File: src/main/java/com/emc/object/s3/request/PutObjectRequest.java [hunk 30]
```diff
@@ -77,6 +85,14 @@ public PutObjectRequest(PutObjectRequest other) {
         if (ifNoneMatch != null) RestUtil.putSingle(headers, RestUtil.HEADER_IF_NONE_MATCH, ifNoneMatch);
         if (acl != null) headers.putAll(acl.toHeaders());
         if (cannedAcl != null) RestUtil.putSingle(headers, S3Constants.AMZ_ACL, cannedAcl.getHeaderValue());
+        if (objectLockLegalHold != null) RestUtil.putSingle(headers, S3Constants.AMZ_OBJECT_LOCK_LEGAL_HOLD, objectLockLegalHold.getStatus());
+        if (objectLockRetention != null) {
+            RestUtil.putSingle(headers, S3Constants.AMZ_OBJECT_LOCK_MODE, objectLockRetention.getMode());
+            if (objectLockRetention.getRetainUntilDate() != null) {
+                RestUtil.putSingle(headers, S3Constants.AMZ_OBJECT_LOCK_RETAIN_UNTIL_DATE,
+                        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").format(objectLockRetention.getRetainUntilDate().toInstant().atOffset(ZoneOffset.UTC)));
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/PutObjectRequest.java]: fixed.

File: src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java [hunk 31]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Override
+    public Map<String, String> getQueryParams() {
+        Map<String, String> queryParams = super.getQueryParams();
+        if (versionId != null) queryParams.put("versionId", versionId);
+        return queryParams;
+    }
+
+    @Override
+    public Map<String, List<Object>> getHeaders() {
```
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java:26]: fixed.

File: src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/SetObjectLegalHoldRequest.java:38]: This request does not go with data, should behave similarly to SetObjectAclRequest where false is returned.

**Review by twincitiesguy — APPROVED** [review]

**Pull request merged**